android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/util/platforms/android/AndroidCallableThread.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.util.platforms.android;
+
+import com.aws.greengrass.logging.api.Logger;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.aws.greengrass.android.component.utils.Constants.EXIT_CODE_FAILED;
+import static com.aws.greengrass.android.component.utils.Constants.EXIT_CODE_SUCCESS;
+import static com.aws.greengrass.android.component.utils.Constants.EXIT_CODE_TERMINATED;
+
+/**
+ * Run callable in thread. Provide Process interface.
+ */
+public class AndroidCallableThread extends Process {
+    private static String COMMAND = "command";
+
+    private final AtomicInteger exitCode = new AtomicInteger(EXIT_CODE_SUCCESS);
+    private final Thread thread;
+
+    /**
+     * Creates AndroidCallableThread instance.
+     *
+     * @param callable callable to run
+     * @param logger component's logger
+     * @param command command to use for logging
+     * @param onExit will run when thread is gone
+     */
+    AndroidCallableThread(Callable<Integer> callable, Logger logger, String[] command, Runnable onExit) {
+        super();
+        thread = new Thread(() -> {
+            try {
+                logger.atDebug().kv(COMMAND, command).log("AndroidCallableThread started");
+                int exitValue = callable.call();
+                exitCode.set(exitValue);
+                logger.atDebug().kv(COMMAND, command).kv("exitValue", exitValue)
+                        .log("AndroidCallableThread finished");
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)